services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/PushStatusCollector.java

```diff
@@ -0,0 +1,194 @@
+        daVinciStoreTopicToPartitionCountMap.put(topicName, new TopicPushStatus(partitionCount));
+      }
+    }
+  }
+
+  public void unsubscribeTopic(String topicName) {
+    daVinciStoreTopicToPartitionCountMap.remove(topicName);
+  }
+
+  public void handleServerPushStatusUpdate(String topicName, ExecutionStatus executionStatus, String detailsString) {
+
+    TopicPushStatus topicPushStatus = daVinciStoreTopicToPartitionCountMap.get(topicName);
+    if (topicPushStatus == null) {
+      if (executionStatus.equals(ExecutionStatus.COMPLETED)) {
+        pushCompletedHandler.accept(topicName);
+      } else if (executionStatus.equals(ExecutionStatus.ERROR)) {
+        pushErrorHandler.accept(topicName, detailsString);
+      }
+    } else {
+      // Update the server topic status in the data structure and wait for async DVC status scan thread to pick up.
+      topicPushStatus.setServerStatus(new ExecutionStatusWithDetails(executionStatus, detailsString));
+    }
+  }
+
+  private void scanDaVinciPushStatus() {
+    for (Map.Entry<String, TopicPushStatus> entry: daVinciStoreTopicToPartitionCountMap.entrySet()) {
+      String topicName = entry.getKey();
+      TopicPushStatus pushStatus = entry.getValue();
+      if (!pushStatus.isMonitoring()) {
+        continue;
+      }
+      ExecutionStatusWithDetails daVinciStatus = null;
+      if (pushStatus.getDaVinciStatus() != null && pushStatus.getDaVinciStatus().getStatus().isTerminal()) {
+        daVinciStatus = pushStatus.getDaVinciStatus();
+      } else {
+        Pair<ExecutionStatus, String> topicDaVinciPushStatus = PushMonitorUtils.getDaVinciPushStatusAndDetails(
+            pushStatusStoreReader,
+            topicName,
+            pushStatus.getPartitionCount(),
+            Optional.empty());
+
+        daVinciStatus =
+            new ExecutionStatusWithDetails(topicDaVinciPushStatus.getFirst(), topicDaVinciPushStatus.getSecond());
+        pushStatus.setDaVinciStatus(daVinciStatus);
+      }
+      ExecutionStatusWithDetails serverStatus = pushStatus.getServerStatus();
+      if (serverStatus == null) {
+        continue;
+      }
+      LOGGER.info(
+          "Topic server push status: {}, Da Vinci push status: {}",
+          serverStatus.getStatus(),
+          daVinciStatus.getStatus());
+      if (isOverallPushCompleted(serverStatus, daVinciStatus)) {
+        pushStatus.setMonitoring(false);
+        pushCompletedHandler.accept(topicName);
+      } else if (isOverallPushError(serverStatus, daVinciStatus)) {
+        pushStatus.setMonitoring(false);
+        String overallErrorString = "";
+        if (serverStatus.getStatus().equals(ExecutionStatus.ERROR)) {
+          overallErrorString = overallErrorString + "Server push error: " + serverStatus.getDetails() + "\n";
+        }
+        if (daVinciStatus.getStatus().equals(ExecutionStatus.ERROR)) {
+          overallErrorString = overallErrorString + "Da Vinci push error: " + daVinciStatus.getDetails() + "\n";
```

Comment:
can we concat the string instead of overwriting the previous status?

Reply:
Hmm Could you give an example here for the conditional concat string?